libs/wingc/src/jsify.rs

```diff
@@ -1233,136 +1223,188 @@ impl<'a> FieldReferenceVisitor<'a> {
+		let mut capture = vec![];
 
-		// don't allow capturing reassignable (`var`) fields.
-		if field_kind.reassignable {
-			self.diagnostics.push(Diagnostic {
-				level: DiagnosticLevel::Error,
-				message: format!(
-					"Unable to reference \"this.{}\" from inflight method \"{}\" because it is reassignable (\"var\")",
-					property.name, self.method_name.name,
-				),
-				span: Some(property.span.clone()),
-			});
+		while index < parts.len() {
+			let curr = parts.get(index).unwrap();
 
-			return true;
-		}
+			let Some(variable) = &curr.variable else {
+				panic!("unexpected - all components should have a variable at this point");
+			};
 
-		// check if the type is capturable (resource, primitive or immutable collection)
-		if !field_kind.type_.is_capturable() {
-			self.diagnostics.push(Diagnostic {
-				level: DiagnosticLevel::Error,
-				message: format!(
-					"Unable to reference \"this.{}\" from inflight method \"{}\" because type {} is not capturable",
-					property.name, self.method_name.name, field_kind.type_,
-				),
-				span: Some(property.span.clone()),
-			});
+			// we have lift off (reached an inflight component)! break our search.
+			if variable.flight == Phase::Inflight {
+				break;
+			}
 
-			return true;
-		}
+			// now we need to verify that the component can be captured.
+			// (1) non-reassignable
+			// (2) capturable type (immutable/resource).
 
-		debug!(
-			"field \"this.{}\" is referenced from inflight method {}",
-			property.name, self.method_name.name
-		);
+			// if the variable is reassignable, bail out
+			if variable.reassignable {
+				self.diagnostics.push(Diagnostic {
+					level: DiagnosticLevel::Error,
+					message: format!("Cannot capture reassignable field '{}'", curr.text),
+					span: Some(curr.expr.span.clone()),
+				});
 
-		// get/create an "ops" set for this field
-		let ops = self.references.entry(property.name.clone()).or_default();
+				return;
+			}
 
-		if !self.path.is_empty() {
-			let op = self.path[0].clone();
+			// if this type is not capturable, bail out
+			if !variable.type_.is_capturable() {
+				self.diagnostics.push(Diagnostic {
+					level: DiagnosticLevel::Error,
+					message: format!(
+						"Cannot capture field '{}' with non-capturable type '{}'",
+						curr.text, variable.type_
+					),
+					span: Some(curr.expr.span.clone()),
+				});
+
+				return;
+			}
 
-			// if we are referencing a resource, verify that the operation is an inflight method
-			if let Some(r) = field_kind.type_.as_resource() {
-				if r
-					.get_method(&op)
-					.filter(|v| v.flight == Phase::Inflight && !v.is_static)
-					.is_none()
-				{
+			// okay, so we have a non-reassignable, capturable type.
+			// one more special case is collections. we currently only support
+			// collections which do not include resources because we cannot
+			// qualify the capture.
+			if let Some(inner_type) = variable.type_.collection_item_type() {
+				if inner_type.as_resource().is_some() {
```

Comment:
This doesn't handle collections of collections of resources - maybe we want some kind of `contains_resources()` function?

(not that important)

Reply:
https://github.com/winglang/wing/issues/1830